multicluster/controllers/multicluster/label_identity_controller.go

```diff
@@ -0,0 +1,303 @@
+	for i, pod := range podList.Items {
+		requests[i] = reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      pod.GetName(),
+				Namespace: pod.GetNamespace(),
+			},
+		}
+	}
+	return requests
+}
+
+// This function needs to be called with labelMutex acquired.
+func (r *LabelIdentityReconciler) setLabelExportUpdatesInProgress(label string) {
+	r.labelExportUpdatesInProgress.Insert(label)
+	// Unlock the labelMutex to allow cache updates for other labels. Any worker operating on
+	// the label will be blocked as it is added into labelExportUpdateInProgress.
+	r.labelMutex.Unlock()
+}
+
+func (r *LabelIdentityReconciler) unsetLabelExportUpdatesInProgress(label string) {
+	r.labelMutex.Lock()
+	r.labelExportUpdatesInProgress.Delete(label)
+}
+
+// removeLabelForPod removes the Pod and label identity mapping from the cache, and deletes label
+// identity ResourceExport if necessary (the Pod update/deletion event causes a label identity no
+// longer present in the cluster).
+// Note that this function will acquire the reconciler's labelMutex, and will keep it locked after
+// it returns. This is useful for eliminating additional context switching when handling Pod
+// update events, since after deleting stale label the labelMutex will immediately need to be held
+// again for checking label add.
+func (r *LabelIdentityReconciler) removeLabelForPod(ctx context.Context, podNamespacedName string) (ctrl.Result, error) {
+	r.labelMutex.Lock()
+	if originalLabel, isCached := r.podLabelCache[podNamespacedName]; isCached {
+		if r.labelExportUpdatesInProgress.Has(originalLabel) {
+			return ctrl.Result{Requeue: true}, nil
+		}
+		// Check if the original label is stale.
+		if podNames, ok := r.labelToPodsCache[originalLabel]; ok && len(podNames) == 1 && podNames.Has(podNamespacedName) {
+			klog.V(2).InfoS("Label no longer exists in the cluster, deleting corresponding ResourceExport", "label", originalLabel)
+			r.setLabelExportUpdatesInProgress(originalLabel)
+			err := r.deleteLabelIdentityResExport(ctx, originalLabel)
+			r.unsetLabelExportUpdatesInProgress(originalLabel)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			r.deleteFromLabelCache(podNamespacedName, originalLabel)
+		} else {
+			// The original label still has other Pod that refers to it. Update the cache with
+			// labelMutex acquired.
+			podNames.Delete(podNamespacedName)
+			delete(r.podLabelCache, podNamespacedName)
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+// onPodCreateOrUpdate updates the Pod and label identity mapping in the cache, and
+// updates label identity ResourceExport if necessary (the Pod creation/update
+// event causes a new label identity to appear in the cluster or a label identity
+// no longer present in the cluster or both).
+func (r *LabelIdentityReconciler) onPodCreateOrUpdate(ctx context.Context, podNamespacedName, currentLabel string) (ctrl.Result, error) {
+	// Based on the reconciler's predicate, the Pod has updated labels.
```

Comment:
Could we make sure on this? The cost may be high to delete and recreate Export, as it leads flow updates on all members.

Reply:
Added `labelChangedPredicate` for Namespace events as well. Also added a quick cache check in `onPodCreateOrUpdate` with mutex acquired only once